OpenRA.Mods.Common/Traits/UpgradesIcons.cs

```diff
@@ -0,0 +1,137 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Attach this to the player actor. When attached, enables all actors possessing the UpgradesWhenCreated ",
+		"trait to have their production queue icons render with an overlay defined in this trait. ",
+		"The icon change occurs when UpgradesWhenCreated.Prerequisites are met.")]
+	public class UpgradesIconsInfo : ITraitInfo, Requires<TechTreeInfo>
+	{
+		[Desc("Image used for this decoration. Defaults to the actor's type.")]
+		public readonly string Image = "rank";
+
+		[Desc("Sequence used for this decoration (can be animated).")]
+		public readonly string Sequence = "rank";
+
+		[Desc("Palette to render the sprite in. Reference the world actor's PaletteFrom* traits.")]
+		public readonly string Palette = "effect";
+
+		[Desc("Point in the actor's bounding box used as reference for offsetting the decoration image." +
+			"Possible values are any combination of Top, VCenter, Bottom and Left, HCenter, Right separated by a comma.")]
+		public readonly ReferencePoints ReferencePoint = ReferencePoints.Top | ReferencePoints.Right;
+
+		[Desc("Pixel offset relative to the actor's bounding box' reference point.")]
+		public readonly int2 Offset = new int2(-7, 6);
+
+		[Desc("Visual scale of the image.")]
+		public readonly float Scale = 2f;
+
+		public object Create(ActorInitializer init) { return new UpgradesIcons(init, this); }
+	}
+
+	public class UpgradesIcons : ITechTreeElement
+	{
+		public readonly Sprite Sprite;
+		public readonly string Palette;
+		public readonly float Scale;
+
+		const string Prefix = "upgradedVariant.";
```

Comment:
Not as far as I know since this is an internally used value and should not be visible outside the code. I know hard-coding is not a desirable practice, but I've just imitated what I've seen in `GlobalUpgradeManager.cs`. 

The problem is, if I remember correctly, that the `TechTree` doesn't associate `Watcher.Key` with the registering `ITechTreeElement`. So in a situation where multiple `ITechTreeElement`s register `Watcher`s with the same `Key`, and one removes its `Watcher`, `Watcher`s for all `ITechTreeElement`s get removed.